### ISSUE TITLE
`Query::each()` key started with 1 instead of 0 (no `indexBy`)

### ISSUE THREAD

**Issue by Vovan-VE** [yiisoft/yii2#14916]
My API in a project produce data with multiple arrays of different descendants of `ActiveRecord`. So I cannot feed just one `DataProviderInterface` to `Json::encode()` (via `Serializer`), because `Json::encode()` itself does not expand `DataProviderInterface` to models under it.

So, I feed to `Json::encode()` array with a number of `Traversable`s somewhere inside, because `Json::encode()` does `foreach` over them.

The problem is that `$query->each()` (without `indexBy`) starts keys with `1` instead of `0`. 
So, `Json::encode()` will produce object `{"1": {...}, "2": {...}, ...}`
instead of array `[{...}, {...}, ...]`

This caused by line https://github.com/yiisoft/yii2/blob/2.0.12/framework/db/BatchQueryResult.php#L121

### What steps will reproduce the problem?

```php
echo Json::encode([
    'foo' => [
        'bar' => Customer::find()
            ->limit(3)
            ->each(),
    ],
]);
```

### What is the expected result?

```json
{"foo":{"bar":[{...},{...},{...}]}}
```

### What do you get instead?

```json
{"foo":{"bar":{"1":{...},"2":{...},"3":{...}}}}
```

### Additional info

| Q                | A
| ---------------- | ---
| Yii version      | 2.0.12
| PHP version      | any (actually 7.0.22)

**Comment by samdark** [comment]
Why do you need `each()` in this case?

**Comment by Vovan-VE** [comment]
I want pass it as `Traversable` to `Json::encode()` somewhere inside. Current solution is generator:

```php
public function getCustomers()
{
    foreach ($query->each() as $model) {
        yield $model;
    }
}

echo Json::encode([
    'foo' => [
        'bar' => $foo->getCustomers(),
    ],
]);
```

**Comment by samdark** [comment]
Can't you pass `$query->all()` instead?

**Comment by Vovan-VE** [comment]
Yes, but... Stop. `Json::encode()` will do it anyway. Ok, there is no need almost for all of this. The only difference is a number of loops performed.

**Comment by samdark** [comment]
OK. The indexing `each()` does seems to be non-natural but I'm afraid we can't easily change it. Need to check if it's covered by any existing tests or mentioned in docs.

**Comment by samdark** [comment]
If we'll do change it, UPGRADE notes should be written.

**Comment by Vovan-VE** [comment]
There are only two test cases: 1) no `indexBy()` and no keys https://github.com/yiisoft/yii2/blob/2.0.13/tests/framework/db/BatchQueryResultTest.php#L83
2) keys with `indexBy()` near there at line 95.

**Comment by klimov-paul** [comment]
Resolved by commit 00afe6a4b8d80c8ee475f831a657f0adad77dbee
